src/bedrockExplainPatch.js

```diff
@@ -0,0 +1,139 @@
+  'anthropic.claude-3-haiku-20240307-v1:0': anthropicCountTokens,
+  'anthropic.claude-3-opus-20240229-v1:0': anthropicCountTokens,
+  'anthropic.claude-instant-v1': anthropicCountTokens,
+  'ai21.j2-mid-v1': null,
+  'ai21.j2-ultra-v1': null,
+  'cohere.command-text-v14': null,
+  'cohere.command-light-text-v14': null,
+  'cohere.embed-english-v3': null,
+  'cohere.embed-multilingual-v3': null,
+  'meta.llama2-13b-chat-v1': null,
+  'meta.llama2-70b-chat-v1': null,
+  'meta.llama3-8b-instruct-v1:0': null,
+  'meta.llama3-70b-instruct-v1:0': null,
+  'mistral.mistral-7b-instruct-v0:2': null,
+  'mistral.mixtral-8x7b-instruct-v0:1': null,
+  'mistral.mistral-large-2402-v1:0': null,
+  'stability.stable-diffusion-xl-v0': null,
+  'stability.stable-diffusion-xl-v1': null
+}
+
+const countTokens = (text, modelId) => {
+  if (COUNT_TOKENS_HASHFUN[modelId]) {
+    return COUNT_TOKENS_HASHFUN[modelId](text)
+  }
+
+  throw new Error(`Model ${modelId} not supported for token counting`)
+}
+
+/* eslint-disable camelcase */
+export default async function explainPatch ({
+  patchBody, owner, repo,
+  models = ['anthropic.claude-3-opus-20240229-v1:0'],
+  system = `
+You are an expert software engineer reviewing a pull request on Github. Lines that start with "+" have been added, lines that start with "-" have been deleted. Use markdown for formatting your review.
+
+Desired format:
+### Description
+<description_of_PR> // How does this PR change the codebase? What is the motivation for this change?
+
+### Changes
+<list_of_changes> // Describe the main changes in the PR, organizing them by filename
+
+### Security Hotspots
+<list_of_security_hotspots> // Describe locations for possible vulnerabilities in the change, order by risk
+\n`,
+  max_tokens = 2048,
+  temperature = 1,
+  top_p = 1,
+  amplification = 2,
+  region = 'us-east-1',
+  debug = false
+}) {
+  const userPrompt = `Repository: https://github.com/${owner}/${repo}\n\nThis is the PR diff\n\`\`\`\n${patchBody}\n\`\`\``
+  const realModels = Array.isArray(models) ? models : models.split(' ')
+
+  if (debug) {
+    console.log(`user_prompt:\n\n${userPrompt}`)
+  }
+
+  const client = new BedrockRuntimeClient({ region })
+
+  let modelId = null
+  let response = null
+
+  // iterate over the models, and find one that works, or throw the last error
+  // catch all the original errors in between, and throw the last one
+  for (let i = 0; i < realModels.length; i++) {
+    try {
+      modelId = realModels[i]
+
+      const pLen = countTokens(patchBody, modelId)
+
+      if (pLen === 0) { throw new Error('The patch is empty, cannot summarize!') }
+      if (pLen < amplification * max_tokens) { throw new Error('The patch is trivial, no need for a summarization') }
+
+      const command = new InvokeModelCommand({
+        modelId,
+        contentType: 'application/json',
+        accept: 'application/json',
+        body: JSON.stringify({
+          anthropic_version: 'bedrock-2023-05-31',
```

Comment:
should this be configurable?

Reply:
the whole `body` argument should be configured per model, it seems. They don't abstract that part in AWS. I'll work on a more generic solution the next iteration, or when we will need to run any of the other model